vm/vm.go

```diff
@@ -591,37 +554,43 @@ func parse(
 	// account will be spawned to the wrong location, but it's much cheaper to perform this check
 	// now and fail fast.
 
-	// in order to calculate the principal, we need to extract the pubkey from the spawn tx
-	var unmarshaled struct {
-		*athcon.MethodSelector
-		signing.PublicKey
+	var payload athcon.Payload
+	err = gossamerScale.Unmarshal(output.Payload, &payload)
+	if err != nil {
+		return nil, nil, fmt.Errorf("%w: malformed TX payload: %w", core.ErrMalformed, err)
 	}
-	err = gossamerScale.Unmarshal(output.Payload, &unmarshaled)
+	args, err := wallet.ParseArgs(payload)
 	if err != nil {
-		return nil, nil, fmt.Errorf("%w: malformed spawn payload", core.ErrMalformed)
+		return nil, nil, fmt.Errorf("%w: parsing TX arguments: %w", core.ErrMalformed, err)
 	}
 
 	// now that the tx has been parsed, we can perform some more sanity checks.
 	// if this is a spawn for an account that was already spawned, we may
 	// have assumed above that it was not a spawn. now we can make sure.
 	// that this will also catch a non-spawn tx with an unspawned principal, which we
 	// provisionally assumed was a spawn tx.
-	if spawnSelector, err := athcon.FromString("athexp_spawn"); err != nil {
-		return nil, nil, fmt.Errorf("%w: failed to create spawn selector: %w", core.ErrInternal, err)
-	} else if principalAccount.TemplateAddress != nil && *unmarshaled.MethodSelector == spawnSelector {
-		return nil, nil, fmt.Errorf("%w: principal account already spawned", core.ErrMalformed)
-	} else if ctx.IsSpawn() && *unmarshaled.MethodSelector != spawnSelector {
-		return nil, nil, fmt.Errorf("%w: non-spawn tx with unspawned principal", core.ErrNotSpawned)
-	}
 
-	computedPrincipal := core.ComputePrincipalFromPubkey(
-		ctx.Header.TemplateAddress,
-		unmarshaled.PublicKey,
-	)
+	switch args := args.(type) {
+	case *wallet.SpawnArgs:
+		if ctx.PrincipalAccount.TemplateAddress != nil {
+			return nil, nil, fmt.Errorf("%w: principal account already spawned", core.ErrMalformed)
+		}
+		ctx.SpawnTx = true
+		ctx.Header.MaxGas = core.ATHENA_GAS_SPAWN + core.ATHENA_GAS_VERIFY
+
+		computedPrincipal := core.ComputePrincipalFromPubkey(ctx.Header.TemplateAddress, args.Pubkey[:])
+		if computedPrincipal != principal {
+			return nil, nil, fmt.Errorf(
+				"%w: calculated spawn principal does not match %s", core.ErrMalformed, principal.String())
```

Comment:
```
"%w: calculated spawn principal %s does not match %s", core.ErrMalformed, computedPrincipal, principal.String()
```

Please also use the same wording in variable name as in error msg. `calculated` or `computed`.